react/components/EditorContainer/Sidebar/ComponentList/utils.ts

```diff
@@ -15,97 +17,97 @@ export const getParentTreePath = (treePath: string): string => {
+export const isRootComponentRelative = (minimumTreePathSize: number) => (
+  rootComponent: SidebarComponent
+) => {
   const splitTreePath = rootComponent.treePath.split('/')
 
   return (
-    splitTreePath[1].startsWith('$') ||
-    (splitTreePath.length === 3 && splitTreePath[2].startsWith('$'))
+    splitTreePath[minimumTreePathSize - 1].startsWith('$') ||
+    (splitTreePath.length === minimumTreePathSize + 1 &&
+      splitTreePath[minimumTreePathSize].startsWith('$'))
   )
 }
 
-export const normalize = (components: SidebarComponent[]) => {
-  const [roots, leaves] = partition(isRootComponent, components)
-
-  const [relativeRoots, nonRelativeRoots] = partition(
-    isRootComponentRelative,
-    roots
+const mountFullTree = (
+  root: SidebarComponent,
+  rootOwnLeaves: SidebarComponent[]
+): NormalizedComponent => {
+  if (rootOwnLeaves.length === 0) {
+    return { ...root, isSortable: false }
+  }
+  const rootTreePathLength = pathLength(root.treePath)
+  const immediateChildren = rootOwnLeaves.filter(
+    ({ treePath: childPath }) =>
+      rootTreePathLength + 1 === pathLength(childPath)
+  )
+  const nonImmediateChildren = rootOwnLeaves.filter(
+    ({ treePath: childPath }) => rootTreePathLength + 1 < pathLength(childPath)
   )
 
-  const normalizedRelativeRoots = relativeRoots.reduce<NormalizedRelativeRoot>(
-    (acc, currRoot) => {
-      const splitTreePath = currRoot.treePath.split('/')
-
-      const treePathTail = splitTreePath[splitTreePath.length - 1]
-
-      const blockRole = getBlockRole(treePathTail)
+  return {
+    ...root,
+    components: immediateChildren.map(child => {
+      const grandChildren = nonImmediateChildren.filter(({ treePath }) =>
+        treePath.startsWith(child.treePath)
+      )
+      return mountFullTree(child, grandChildren)
+    }),
+    isSortable: false,
+  }
+}
 
-      if (!blockRole) {
-        return acc
-      }
+export const normalize = (components: SidebarComponent[]) => {
+  const minimumTreePathSize = components.reduce((acc, currentComponent) => {
+    const treePathSize = currentComponent.treePath.split('/').length
+    return treePathSize < acc ? treePathSize : acc
+  }, Infinity)
+
+  const [roots, leaves] = partition(
+    isRootComponent(minimumTreePathSize),
+    components
+  )
 
-      const currNormalizedRoot = { ...currRoot, isSortable: false }
+  const reduceRoot = (
+    acc: SidebarComponent[],
+    root: SidebarComponent
+  ): SidebarComponent[] => {
+    const { treePath: currentRootPath } = root
+    const rootPathLength = pathLength(currentRootPath)
+    const childrenFromRoot = leaves.filter(
+      ({ treePath: leafTreePath }) =>
+        leafTreePath.startsWith(currentRootPath) &&
+        rootPathLength + 1 === pathLength(leafTreePath)
+    )
 
-      if (blockRole === 'around' && splitTreePath.length === 3) {
-        return {
-          ...acc,
-          nestedAround: [...acc.nestedAround, currNormalizedRoot],
-        }
-      }
+    const rootObject = {
```

Comment:
This could be moved to an external function that could also be used [here](https://github.com/vtex-apps/admin-pages/pull/194/commits/9b4ac5a781a19f5ed60077dfbbb42d4a7bb3117d#diff-125db75cb1075f422912f53d021a0443R60).